### PR TITLE
WIP Build two arch version 2

### LINKED ISSUE #62: Build for node containing CPU+GPU

### PR DESCRIPTION
**Pull request by tpatki**
# Description

Rebasing Stephanie's existing PR #121 was hard due to several conflicts and changes in variorum, so this PR is a fresh start at adding her changes on v0.6.0, and then further add support for AMD CPU, GPU, Intel GPU, JSON, docs, etc.. 

Fixes #62.
Supersedes #121 .

ToDo:

- [x] Add AMD CPU support
- [x] Fix AMD build error with DENABLE_WARNINGS=On (submit separate PR)
- [x] Update JSON APIs to be node-level and not component level (similar to best effort node power capping), node-level APIs are not supported on GPU-only builds. The GPU values should be populated internally through a different GPU-specific API at some point in a different PR. 

~~- [ ] Update JSON APIs to support multiple-platforms~~ <- separate PR as we need to add JSON API for GPUs so it can populate the JSON string at node level. 

- [x] Update the best_effort-node-power-limit code to maintain its v0.6.0 release behavior, it is a node-level API not a component-level (CPU, Mem, GPU level) API. 
- [x] Set a MAX_PLATFORMS variable so we can easily update in the future if we have a situation where we are building for a CPU + GPU + Other Accelerator situation.

~- [ ] Update documentation to reflect how to build for CPU + GPU, and how we only support 2 platforms at a time~ <-- separate docs PR 

- [x] (Start a separate discussion about how to handle component power capping and what "best effort" means when GPU power capping is available)

## Type of change

- [x] New feature/architecture support (non-breaking change which adds functionality)
- [ ] Documentation update
- [x] Build/CI update

# How Has This Been Tested?

Tested build with -DENABLE_WARNINGS=On,  as well as variorum-print-power-example, variorum-cap-best-effort-node-power-cap-example and variorum-get-node-power-json-example wherever applicable.

_Note that the best-effort-power-cap and get-node-power-json examples are node-level, and will only work on the primary platform component which we assume to currently be the CPU. Future goal includes adding support for obtaining detailed GPU information for the node-level APIs -- current behavior of variorum is maintained in this PR for these node-level functions.._ 

- [x] Comus single platform
- [x] Rhetoric single platform
- [x] Thompson single platform
- [x] Juno single platform
- [x] Corona GPU single platform
- [x] Lassen CPU single platform
- [x] Lassen GPU single platform
- [x] Lassen CPU + GPU (two platforms): print power, node-level json and best effort cap works as expected. Best effort example output is confusing as when printing, NVML power limit doesn't show as changed even though it has (a derived power cap is indirectly chosen by IBM to set through IBM's node limit, which supersedes any NVML limits.).
- [x] Tioga CPU single platform
- [x] Tioga GPU single platform
- [x] Tioga CPU+GPU single platform: Similar to Lassen, print power and print power limit and JSON APIs work as expected. Best effort power cap cannot be tested here due to lack of permissions. 

# Checklist:

- [x] I have run `./scripts/check-code-format.sh` and confirm my C/C++ code follows the style guidelines of variorum
- [ ] I have run `flake8` and `black --check --diff` and confirm my python code follows the style guidelines of variorum
- [ ] I have run `./scripts/check-rst-format.sh` and confirm my documentation follows the style guidelines of variorum
- [x] I have added comments in my code
- [x] My changes generate no new warnings (build with `-DENABLE_WARNINGS=ON`)
- [x] New and existing unit tests pass with my changes

Thank you for taking the time to contribute to Variorum!

### CODE REVIEW

File: src/variorum/variorum.h [hunk 1]
```diff
@@ -226,7 +226,6 @@ int variorum_print_verbose_thermals(void);
 /// - Intel Kaby Lake
 /// - Intel Cascade Lake
 /// - Intel Cooper Lake
-/// - Intel Arctic Sound
```
Comment (slabasan) [inline src/variorum/variorum.h]: Why are these being removed?

File: src/variorum/config_architecture.h [hunk 2]
```diff
@@ -268,7 +282,8 @@ int variorum_exit(const char *filename,
 
 void variorum_get_topology(unsigned *nsockets,
                            unsigned *ncores,
-                           unsigned *nthreads);
+                           unsigned *nthreads,
+                           int idx);
```
Comment (slabasan) [inline src/variorum/config_architecture.h:286]: what is this index being used for?

**Review by slabasan — DISMISSED** [review]
Build looks good to me, will you be adding a host config file for the tioga dual build platform as well?

File: src/variorum/config_architecture.h (hunk 2 above)
Comment (tpatki) [reply src/variorum/config_architecture.h:286]: The number of sockets, threads etc and the hostname is set in the variorum_get_topology function in config_architecture.c, which still needs to be done per platform.

File: src/variorum/variorum.h (hunk 1 above)
Comment (tpatki) [reply src/variorum/variorum.h]: That looks like a mistake, let me revert that shortly.

File: src/variorum/config_architecture.h (hunk 2 above)
Comment (slabasan) [reply src/variorum/config_architecture.h:286]: I'm seeing that sometimes `P_MSR_CORE_IDX` is being passed into `variorum_get_topology`, what does this index represent?

File: src/variorum/variorum.h (hunk 1 above)
Comment (tpatki) [reply src/variorum/variorum.h]: Actually, not a mistake, @slabasan . Intel Arctic Sound is a GPU, and doesn't support these APIs (verbose ones) in the port.

**Comment by tpatki** [comment]
@slabasan Added Tioga host configs to a separate PR.

File: src/variorum/config_architecture.h (hunk 2 above)
Comment (tpatki) [reply src/variorum/config_architecture.h:286]: `MSR_CORE_IDX` was added for scenarios where we need to share the `msr_core.c` functionality -- especially among Intel and AMD CPUs. Previously, `msr_core.c` was using the `P_INTEL_CPU_IDX` internally, which will not work correctly if we have an AMD CPU platform, which would need the `P_AMD_CPU_IDX` instead. So, I added the `MSR_CORE_IDX`, which can now be set based on the platform that we're building for, and be passed to common internal interfaces such as the ones from `msr_core.c`.

**Comment by tpatki** [comment]
@slabasan : as discussed, reverted the change around Arctic Sound (will submit separate PR), and removed the `<Inte/AMD> is idx` prints, and moved the number of registered platforms print to the variorum log instead. Let me know if I missed anything, I believe this is ready to go. Thank you!

**Review by slabasan — APPROVED** [review]

**Pull request merged**